leak_detector/leak_detector.py

```diff
@@ -0,0 +1,79 @@
+"""
+Find memory leaks when malloc and free are used.
+Starting from the current position it matches all calls to malloc
+with calls to free with the same pointer and
+keeps track of the calls that have no corresponding call to free.
+It is recommended to go to the start of time and then use the command.
+It prints a full list of unmatched calls at the end.
+    Usage: mleak
+
+Contibutors:  Emiliano Testa
+Copyright (C) 2022 Undo Ltd
+"""
+
+import copy
+import gdb
+
+from undodb.debugger_extensions import (
+    udb,
+)
+
+
+ALLOC_FN = "malloc"
+FREE_FN = "free"
+all_allocs = []
+
+
+class MemAlloc:
+    def __init__(self, addr, size, bbcount):
+        self.addr = addr
+        self.size = size
+        self.bbcount = bbcount
+
+
+def handle_alloc_fn():
+    frame = gdb.selected_frame()
+    size = frame.read_register("rdi")
+    bbcount = udb.time.get().bbcount
+    gdb.execute("finish")
+    frame = gdb.selected_frame()
+    addr = frame.read_register("rax")
+    all_allocs.append(MemAlloc(addr, size, bbcount))
+
+
+def handle_free_fn():
+    frame = gdb.selected_frame()
+    addr = frame.read_register("rdi")
+    for alloc in copy.copy(all_allocs):
+        if alloc.addr == addr:
+            all_allocs.remove(alloc)
+
+
+def handle_bp_event(event):
```

Comment:
I thought about using this approach but the documentation states:

```
You should not alter the execution state of the inferior (i.e., step, next, etc.), alter the current frame context (i.e., change the current active frame), or alter, add or delete any breakpoint. As a general rule, you should not alter any data within GDB or the inferior at this time. 
```

and, because I need to get to the end of malloc() I decided not to use this technique and I went for the stop handlers instead.